File: NationalSecurityCommunity/app/src/main/java/com/national/security/community/base/BasePresenter.java
package com.national.security.community.base;

import com.national.security.community.mvp.IPresenter;
import com.national.security.community.mvp.IView;
import com.trello.rxlifecycle2.components.support.RxAppCompatActivity;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * @ description:  Presenter基类
 * @ author:  ljn
 * @ time:  2018/1/2
 */
public class BasePresenter<T extends IView> implements IPresenter<T> {

    protected RxAppCompatActivity mActivity;
    protected T mView;
    private CompositeDisposable mCompositeDisposable;

    public BasePresenter(RxAppCompatActivity mActivity) {
        this.mActivity = mActivity;
    }

    @Override
    public void attachView(T view) {
        this.mView = view;
    }

    @Override
    public void detachView() {
        this.mView = null;
        unSubscribe();
    }


    private void unSubscribe() {
        if (mCompositeDisposable != null) {
            mCompositeDisposable.clear();
        }
    }

    protected void addSubscribe(Disposable disposable) {
        if (mCompositeDisposable == null) {
            mCompositeDisposable = new CompositeDisposable();
        }
        mCompositeDisposable.add(disposable);
    }
}

File: app/src/main/java/com/jmd_news_kotlin/base/BasePresenter.kt
package com.jmd_news_kotlin.base

import RxBus
import RxUtil
import com.jmd_news_kotlin.mvp.IPresenter
import com.jmd_news_kotlin.mvp.IView
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * Created by asus on 2017/6/3.
 */
open class BasePresenter<T : IView> : IPresenter {


    protected var mCompositeDisposable: CompositeDisposable? = null
    protected var mView: T? = null
    override fun detachView() {
        this.mView = null
        unSubscribe()
    }

    protected fun unSubscribe() {
        mCompositeDisposable?.clear()

    }

    protected fun addSubscribe(disposable: Disposable) {
        mCompositeDisposable ?: let { mCompositeDisposable = CompositeDisposable() }
        mCompositeDisposable?.add(disposable)
    }

    protected fun <U> addRxBusSubscribe(eventType: Class<U>) {
        mCompositeDisposable ?: let { mCompositeDisposable = CompositeDisposable() }
        mCompositeDisposable?.add(RxBus.toObservable(eventType).compose(RxUtil.rxSchedulerHelper()).subscribe())
    }

}